cirq/sim/density_matrix_simulator.py

```diff
@@ -256,7 +256,7 @@ def _base_iterator(self,
         initial_matrix = qis.to_valid_density_matrix(initial_state,
                                                      len(qid_shape),
                                                      qid_shape=qid_shape,
-                                                     dtype=self._dtype)
+                                                     dtype=self._dtype).copy()
```

Comment:
This will make an unnecessary copy twice if to_valid_density_matrix already made a copy to change the dtype.

Reply:
It looks like `to_valid_density_matrix` only changes the dtype if `initial_state` is an int, in which case a double-copy isn't particularly expensive. For matrices, a mismatched dtype results in an error instead of copying the matrix.

I think this `copy` should be OK.

Reply:
I think there are other cases too, and it could change in the future.  An easy solution would be to add
```python
if np.may_share_memory(initial_matrix, initial_state):
    initial_matrix = initial_matrix.copy()

Reply:
I move the copy as an optional parameter in `to_valid_density_matrix` to handle this.